File: src-tauri/src/models/database_structure.rs
use serde::{Deserialize, Serialize};

/// データベース構造全体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseStructure {
    pub connection_id: String,
    pub database_name: String,
    pub database_type: String,
    pub schemas: Vec<Schema>,
    pub fetched_at: String,
}

/// スキーマ情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Schema {
    pub name: String,
    pub is_system: bool,
    pub tables: Vec<Table>,
    pub views: Vec<View>,
}

/// テーブル情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Table {
    pub name: String,
    pub schema: String,
    pub comment: Option<String>,
    pub estimated_row_count: Option<i64>,
    pub columns: Vec<Column>,
    pub primary_key: Option<PrimaryKey>,
    pub indexes: Vec<Index>,
    pub foreign_keys: Vec<ForeignKey>,
    pub referenced_by: Vec<ForeignKeyReference>,
}

/// ビュー情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct View {
    pub name: String,
    pub schema: String,
    pub comment: Option<String>,
    pub columns: Vec<Column>,
    pub definition: Option<String>,
}

/// カラム情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Column {
    pub name: String,
    pub data_type: String,
    pub display_type: String,
    pub nullable: bool,
    pub default_value: Option<String>,
    pub is_primary_key: bool,
    pub is_foreign_key: bool,
    pub is_unique: bool,
    pub is_auto_increment: bool,
    pub ordinal_position: i32,
    pub comment: Option<String>,
}

/// プライマリキー情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrimaryKey {
    pub name: String,
    pub columns: Vec<String>,
}

/// インデックス情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Index {
    pub name: String,
    pub is_unique: bool,
    pub is_primary: bool,
    pub columns: Vec<String>,
    #[serde(rename = "type")]
    pub index_type: String,
}

/// 外部キー情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForeignKey {
    pub name: String,
    pub columns: Vec<String>,
    pub referenced_schema: String,
    pub referenced_table: String,
    pub referenced_columns: Vec<String>,
    pub on_delete: String,
    pub on_update: String,
}

/// 外部キー参照情報
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForeignKeyReference {
    pub source_schema: String,
    pub source_table: String,
    pub source_columns: Vec<String>,
    pub target_columns: Vec<String>,
    pub constraint_name: String,
}

File: app/types/database-structure.ts
/**
 * データベース構造全体
 */
export interface DatabaseStructure {
  /** 接続ID */
  connectionId: string;
  /** データベース名 */
  databaseName: string;
  /** データベース種別 */
  databaseType: 'postgresql' | 'mysql' | 'sqlite';
  /** スキーマ一覧 */
  schemas: Schema[];
  /** 取得日時 */
  fetchedAt: string;
}

/**
 * データベース構造サマリー（軽量）
 */
export interface DatabaseStructureSummary {
  /** 接続ID */
  connectionId: string;
  /** データベース名 */
  databaseName: string;
  /** データベース種別 */
  databaseType: 'postgresql' | 'mysql' | 'sqlite';
  /** スキーマ一覧 */
  schemas: SchemaSummary[];
  /** 取得日時 */
  fetchedAt: string;
}

/**
 * スキーマ情報（サマリー）
 */
export interface SchemaSummary {
  /** スキーマ名 */
  name: string;
  /** システムスキーマかどうか */
  isSystem: boolean;
  /** テーブル一覧 */
  tables: TableSummary[];
  /** ビュー一覧 */
  views: TableSummary[];
}

/**
 * テーブル情報（サマリー）
 */
export interface TableSummary {
  /** テーブル名 */
  name: string;
  /** スキーマ名 */
  schema: string;
  /** コメント/説明 */
  comment: string | null;
  /** 推定行数 */
  estimatedRowCount: number | null;
}

/**
 * スキーマ情報
 */
export interface Schema {
  /** スキーマ名 */
  name: string;
  /** システムスキーマかどうか */
  isSystem: boolean;
  /** テーブル一覧 */
  tables: Table[];
  /** ビュー一覧 */
  views: View[];
}

/**
 * テーブル情報
 */
export interface Table {
  /** テーブル名 */
  name: string;
  /** スキーマ名 */
  schema: string;
  /** コメント/説明 */
  comment: string | null;
  /** 推定行数 */
  estimatedRowCount: number | null;
  /** カラム一覧 */
  columns: Column[];
  /** プライマリキー */
  primaryKey: PrimaryKey | null;
  /** インデックス一覧 */
  indexes: Index[];
  /** 外部キー一覧 */
  foreignKeys: ForeignKey[];
  /** 参照されている外部キー */
  referencedBy: ForeignKeyReference[];
}

/**
 * ビュー情報
 */
export interface View {
  /** ビュー名 */
  name: string;
  /** スキーマ名 */
  schema: string;
  /** コメント/説明 */
  comment: string | null;
  /** カラム一覧 */
  columns: Column[];
  /** ビュー定義SQL(オプション) */
  definition: string | null;
}

/**
 * カラム情報
 */
export interface Column {
  /** カラム名 */
  name: string;
  /** データ型 */
  dataType: string;
  /** 表示用データ型(長さ等含む) */
  displayType: string;
  /** NULL許可 */
  nullable: boolean;
  /** デフォルト値 */
  defaultValue: string | null;
  /** プライマリキーの一部か */
  isPrimaryKey: boolean;
  /** 外部キーの一部か */
  isForeignKey: boolean;
  /** ユニーク制約があるか */
  isUnique: boolean;
  /** 自動インクリメントか */
  isAutoIncrement: boolean;
  /** カラム位置(順序) */
  ordinalPosition: number;
  /** コメント/説明 */
  comment: string | null;
}

/**
 * プライマリキー情報
 */
export interface PrimaryKey {
  /** 制約名 */
  name: string;
  /** カラム名リスト */
  columns: string[];
}

/**
 * インデックス情報
 */
export interface Index {
  /** インデックス名 */
  name: string;
  /** ユニークインデックスか */
  isUnique: boolean;
  /** プライマリキーインデックスか */
  isPrimary: boolean;
  /** カラム名リスト */
  columns: string[];
  /** インデックス種別 */
  type: string;
}

/**
 * 外部キー情報
 */
export interface ForeignKey {
  /** 制約名 */
  name: string;
  /** 元カラム */
  columns: string[];
  /** 参照先スキーマ */
  referencedSchema: string;
  /** 参照先テーブル */
  referencedTable: string;
  /** 参照先カラム */
  referencedColumns: string[];
  /** ON DELETE アクション */
  onDelete: string;
  /** ON UPDATE アクション */
  onUpdate: string;
}

/**
 * 外部キー参照情報(このテーブルを参照している外部キー)
 */
export interface ForeignKeyReference {
  /** 参照元スキーマ */
  sourceSchema: string;
  /** 参照元テーブル */
  sourceTable: string;
  /** 参照元カラム */
  sourceColumns: string[];
  /** このテーブルのカラム */
  targetColumns: string[];
  /** 制約名 */
  constraintName: string;
}
